compiler/rustc_next_trait_solver/src/solve/normalizes_to/projection.rs

```diff
@@ -0,0 +1,167 @@
+            // because the former will also record a built-in candidate in the inspector.
+            self.forced_ambiguity(MaybeCause::Ambiguity).map(|cand| cand.result)
+        }
+    }
+
+    #[instrument(level = "debug", skip(self), ret)]
+    fn assemble_and_merge_env_candidates(
+        &mut self,
+        goal: Goal<I, ty::NormalizesTo<I>>,
+    ) -> Option<QueryResult<I>> {
+        // Even when a trait bound has been proven using a where-bound, we
+        // still need to consider alias-bounds for normalization, see
+        // `tests/ui/next-solver/alias-bound-shadowed-by-env.rs`.
+        let (mut candidates, _) =
+            self.assemble_and_evaluate_candidates(goal, AssembleCandidatesFrom::EnvAndBounds);
+        debug!(?candidates);
+
+        if candidates.is_empty() {
+            return None;
+        }
+
+        // If we're normalizing an GAT, we bail if using a where-bound would constrain
+        // its generic arguments.
+        // FIXME(generic_associated_types): Addresses aggressive inference in #92917.
+        //
+        // If this type is a GAT with currently unconstrained arguments, we do not
+        // want to normalize it via a candidate which only applies for a specific
+        // instantiation. We could otherwise keep the GAT as rigid and succeed this way.
+        // See tests/ui/generic-associated-types/no-incomplete-gat-arg-inference.rs.
+        //
+        // This only avoids normalization if a GAT argument is fully unconstrained.
+        // This is quite arbitrary but fixing it causes some ambiguity, see #125196.
+        for arg in goal.predicate.alias.own_args(self.cx()).iter() {
+            let Some(term) = arg.as_term() else {
+                continue;
+            };
+            match self.structurally_normalize_term(goal.param_env, term) {
+                Ok(term) => {
+                    if term.is_infer() {
+                        return Some(self.evaluate_added_goals_and_make_canonical_response(
+                            Certainty::AMBIGUOUS,
+                        ));
+                    }
+                }
+                Err(NoSolution) => return Some(Err(NoSolution)),
+            }
+        }
+
+        // We still need to prefer where-bounds over alias-bounds however.
+        // See `tests/ui/winnowing/norm-where-bound-gt-alias-bound.rs`.
+        if candidates.iter().any(|c| matches!(c.source, CandidateSource::ParamEnv(_))) {
+            candidates.retain(|c| matches!(c.source, CandidateSource::ParamEnv(_)));
+        }
+
+        let result = if let Some((response, _)) = self.try_merge_candidates(&candidates) {
+            Ok(response)
+        } else {
+            self.flounder(&candidates)
+        };
+        Some(result)
+    }
+
+    #[instrument(level = "debug", skip(self), ret)]
+    fn assemble_and_merge_impl_candidates(
+        &mut self,
+        goal: Goal<I, ty::NormalizesTo<I>>,
+    ) -> QueryResult<I> {
+        let (mut candidates, _) =
```

Comment:
comment: something something we already looked for param_env and alias-bound candidates so we don't have to assemble them again